lib/ts/recipe/accountlinking/recipe.ts

```diff
@@ -131,184 +134,208 @@ export default class Recipe extends RecipeModule {
-            let linkAccountsResult = await this.recipeInterfaceImpl.linkAccounts({
-                recipeUserId: user.loginMethods[0].recipeUserId,
-                primaryUserId: primaryUser.id,
-                userContext,
+        // If we got here, we have a session, so we load the session user info.
+        let sessionUser = await getUser(session.getUserId(), userContext);
+        if (sessionUser === undefined) {
+            throw new SessionError({
+                type: SessionError.UNAUTHORISED,
+                message: "Session user not found",
             });
+        }
 
-            if (linkAccountsResult.status === "OK") {
-                logDebugMessage("createPrimaryUserIdOrLinkAccounts successfully linked");
-                return linkAccountsResult.user;
-            } else if (
-                linkAccountsResult.status === "RECIPE_USER_ID_ALREADY_LINKED_WITH_ANOTHER_PRIMARY_USER_ID_ERROR"
-            ) {
-                // this can happen cause of a race condition
-                // wherein the recipe user ID get's linked to
-                // some other primary user whilst this function is running.
-                // But this is OK, we can just return the primary user it is linked to
-                logDebugMessage("createPrimaryUserIdOrLinkAccounts already linked to another user");
-                return linkAccountsResult.user;
-            } else if (linkAccountsResult.status === "INPUT_USER_IS_NOT_A_PRIMARY_USER") {
-                logDebugMessage("createPrimaryUserIdOrLinkAccounts linking failed because of a race condition");
-                // this can be possible during a race condition wherein the primary user
-                // that we fetched somehow is no more a primary user. This can happen if
-                // the unlink function was called in parallel on that user. So we can just retry
-                return await this.createPrimaryUserIdOrLinkAccounts({
-                    tenantId,
-                    user,
+        // If the session user is already primary we re-check if it still isn't linked to the input user
+        if (sessionUser.isPrimaryUser) {
+            logDebugMessage("createPrimaryUserIdOrLinkAccounts session user already primary");
+            // They could become linked in some race conditions
+            if (this.isLinked(sessionUser, inputUser)) {
+                return { status: "OK", user: sessionUser };
+            }
+        } else {
+            // If the session user isn't primary we try and make it one.
+            logDebugMessage("createPrimaryUserIdOrLinkAccounts trying to make session user primary");
+
+            // We could check here if the session user can even become a primary user, but that'd only mean one extra core call
+            // without any added benefits, since the core already checks all pre-conditions
+            if (await this.shouldBecomePrimaryUser(sessionUser, tenantId, session, userContext)) {
```

Comment:
this can also be false cause the verification status does not match the requiresVerification from the shouldDoAutomaticAccountLinking function.. in which case we want to throw an email verification claim validation error..

I would suggest that we remove all these new helper functions and just use the ones from before. It's less error prone that way